expr/tagquery/expression.go

```diff
@@ -45,134 +40,346 @@ func (e Expressions) Strings() []string {
+	// Example
+	// metric1 has tags ["name=a.b.c", "some=value"] in the metric tag index, we evaluate the expression
+	// "anothertag!=value":
+	// 1) expression looks at the metric tag index and it sees that metric1 does not have a tag "anothertag"
+	//    with the value "value", but at this point it doesn't know if another index that will be looked
+	//    at later does, so it returns the decision "none".
+	// 2) expression now looks at index2 and sees again that metric1 does not have the tag and value
+	//    it is looking for, so it returns "none" again.
+	// 3) the expression execution sees that there are no more indexes left, so it applies the default
+	//    decision for the operator != which is "pass", meaning the expression "anothertag!=value" has
+	//    not filtered the metric metric1 out of the result set.
+	//
+	// metric2 has tags ["name=a.b.c", "anothertag=value"] according to the metric tag index and it has
+	// no meta tags, we still evaluate the same expression:
+	// 1) expression looks at metric tag index and see metric2 has tag "anothertag" with value "value".
+	//    it directly comes to a conclusive decision that this metric needs to be filtered out of the
+	//    result set and returns the filter decision "fail".
+	//
+	// metric3 has tags ["name=aaa", "abc=cba"] according to the metric tag index and there is a meta
+	// record assigning the tag "anothertag=value" to metrics matching that query expression "abc=cba".
+	// 1) expression looks at metric3 and sees it does not have the tag & value it's looking for, so
+	//    it returns the filter decision "none" because it cannot know for sure whether another index
+	//    will assign "anothertag=value" to metric3.
+	// 2) expression looks at the meta tag index and it sees that there are meta records matching the
+	//    tag "anothertag" and the value "value", so it retrieves the according filter functions of
+	//    of these meta records and passes metric3's tag set into them.
+	// 3) the filter function of the meta record for the query set "abc=cba" returns true, indicating
+	//    that its meta tag gets applied to metric3.
+	// 4) based on that the tag expression comes to the decision that metric3 should not be part of
+	//    final result set, so it returns "fail".
+	GetDefaultDecision() FilterDecision
+
+	// GetKey returns tag to who's values this expression get's applied to
+	// example:
+	// in the expression "tag1=value" GetKey() would return "tag1"
+	GetKey() string
+
+	// GetValue returns the value part of the expression
+	// example:
+	// in the expression "abc!=cba" this would return "cba"
+	GetValue() string
+
+	// GetOperator returns the operator of this expression
+	GetOperator() ExpressionOperator
+
+	// GetOperatorCost returns a value which should roughly reflect the cost of this operator compared
+	// to other operators. F.e. = is cheaper than =~. Keep in mind that this is only a very rough
+	// estimate and will never be accurate.
+	GetOperatorCost() uint32
```

Comment:
Is there a way that we could run benchmarks on different operators to better determine their cost and fine tune these? Or possibly add metrics about how long it takes to process certain operators so in the future we could do this?

Reply:
Yeah, but the exact number isn't really very relevant as it's only used to compare the expression types against each other. I have done a bunch of benchmarking and tried to tune it as good as possible, which resulted in this commit: https://github.com/grafana/metrictank/pull/1373/commits/bc6c8b7748a7f6a4c35707b79f82ff5bd9dc2785
Basically the sorting by cost first orders by expression type, according to the expressions "operator costs", and then among equal costs it sorts by cardinality of the affected parts of the index